File: people/thom/test/opsm.f
SUBROUTINE ADD_SUB1(f90_in1, f90_in2, f90_in3, f90_out)
  REAL, DIMENSION (0:) :: f90_in1
  REAL, DIMENSION (0:) :: f90_in2
  REAL, DIMENSION (0:) :: f90_in3
  REAL, DIMENSION (0:) :: f90_out
  INTEGER :: shift
  INTEGER :: offset0
  INTEGER :: i0
  INTEGER :: i1
  INTEGER :: i2

! MOA
! out=in1
!

  shift=0* INT (((2)))+0* INT (((3)))+0
  offset0=0* INT (2)+0* INT (6)+0
  DO i0 = 0, ((4))-1
    DO i1 = 0, ((3))-1
      DO i2 = 0, ((2))-1
        f90_out(shift)=f90_in1(offset0)
        offset0=offset0+(1)
        shift=shift+(1)
      END DO
    END DO
  END DO


! MOA
! out=out +  + in2
!

  shift=0* INT (((2)))+0* INT (((3)))+0
  offset0=0* INT (2)+0* INT (6)+0
  DO i0 = 0, ((4))-1
    DO i1 = 0, ((3))-1
      DO i2 = 0, ((2))-1
        f90_out(shift)=f90_out(shift)+f90_in2(offset0)
        offset0=offset0+(1)
        shift=shift+(1)
      END DO
    END DO
  END DO


! MOA
! out=out +  + in3
!

  shift=0* INT (((2)))+0* INT (((3)))+0
  offset0=0* INT (2)+0* INT (6)+0
  DO i0 = 0, ((4))-1
    DO i1 = 0, ((3))-1
      DO i2 = 0, ((2))-1
        f90_out(shift)=f90_out(shift)+f90_in3(offset0)
        offset0=offset0+(1)
        shift=shift+(1)
      END DO
    END DO
  END DO


END SUBROUTINE ADD_SUB1


SUBROUTINE ADD_SUB2(f90_in1, f90_in2, f90_in3, f90_out)
  REAL, DIMENSION (0:) :: f90_in1
  REAL, DIMENSION (0:) :: f90_in2
  REAL, DIMENSION (0:) :: f90_in3
  REAL, DIMENSION (0:) :: f90_out
  INTEGER, DIMENSION (0:2*3*4-1) :: f90_temp
  INTEGER :: shift
  INTEGER :: offset0
  INTEGER :: i0
  INTEGER :: i1
  INTEGER :: i2

! MOA
! temp=in1
!

  shift=0* INT (((2)))+0* INT (((3)))+0
  offset0=0* INT (2)+0* INT (6)+0
  DO i0 = 0, ((4))-1
    DO i1 = 0, ((3))-1
      DO i2 = 0, ((2))-1
        f90_temp(shift)=f90_in1(offset0)
        offset0=offset0+(1)
        shift=shift+(1)
      END DO
    END DO
  END DO


! MOA
! temp=temp * in2
!

  shift=0* INT (((2)))+0* INT (((3)))+0
  offset0=0* INT (2)+0* INT (6)+0
  DO i0 = 0, ((4))-1
    DO i1 = 0, ((3))-1
      DO i2 = 0, ((2))-1
        f90_temp(shift)=f90_temp(shift)*f90_in2(offset0)
        offset0=offset0+(1)
        shift=shift+(1)
      END DO
    END DO
  END DO


! MOA
! temp=temp - in3
!

  shift=0* INT (((2)))+0* INT (((3)))+0
  offset0=0* INT (2)+0* INT (6)+0
  DO i0 = 0, ((4))-1
    DO i1 = 0, ((3))-1
      DO i2 = 0, ((2))-1
        f90_temp(shift)=f90_temp(shift)-f90_in3(offset0)
        offset0=offset0+(1)
        shift=shift+(1)
      END DO
    END DO
  END DO


! MOA
! out=temp
!

  shift=0* INT (((2)))+0* INT (((3)))+0
  offset0=0* INT (2)+0* INT (6)+0
  DO i0 = 0, ((4))-1
    DO i1 = 0, ((3))-1
      DO i2 = 0, ((2))-1
        f90_out(shift)=f90_temp(offset0)
        offset0=offset0+(1)
        shift=shift+(1)
      END DO
    END DO
  END DO


END SUBROUTINE ADD_SUB2


SUBROUTINE ADD_SUB3(f90_in1, f90_in2, f90_in3, f90_out)
  REAL, DIMENSION (0:) :: f90_in1
  REAL, DIMENSION (0:) :: f90_in2
  REAL, DIMENSION (0:) :: f90_in3
  REAL, DIMENSION (0:) :: f90_out
  INTEGER, DIMENSION (0:2*3*4-1) :: f90_temp
  INTEGER :: shift
  INTEGER :: offset0
  INTEGER :: offset1
  INTEGER :: offset2
  INTEGER :: offset3
  INTEGER :: offset4
  INTEGER :: offset5
  INTEGER :: offset6
  INTEGER :: i0
  INTEGER :: i1
  INTEGER :: i2

! MOA
! temp=in2 * in3
!

  shift=0* INT (((2)))+0* INT (((3)))+0
  offset0=0* INT (2)+0* INT (6)+0
  offset1=0* INT (2)+0* INT (6)+0
  DO i0 = 0, ((4))-1
    DO i1 = 0, ((3))-1
      DO i2 = 0, ((2))-1
        f90_temp(shift)=f90_in2(offset0)
        f90_temp(shift)=f90_temp(shift)*f90_in3(offset1)
        offset1=offset1+(1)
        offset0=offset0+(1)
        shift=shift+(1)
      END DO
    END DO
  END DO


! MOA
! out=in1 - four * in3 +  + tfive * temp / in3 - in2
!

  shift=0* INT (((2)))+0* INT (((3)))+0
  offset0=0* INT (2)+0* INT (6)+0
  offset2=0* INT (2)+0* INT (6)+0
  offset4=0* INT (2)+0* INT (6)+0
  offset5=0* INT (2)+0* INT (6)+0
  offset6=0* INT (2)+0* INT (6)+0
  DO i0 = 0, ((4))-1
    DO i1 = 0, ((3))-1
      DO i2 = 0, ((2))-1
        f90_out(shift)=f90_in1(offset0)
        f90_out(shift)=f90_out(shift)-4.000000
        f90_out(shift)=f90_out(shift)*f90_in3(offset2)
        f90_out(shift)=f90_out(shift)+25.000000
        f90_out(shift)=f90_out(shift)*f90_temp(offset4)
        f90_out(shift)=f90_out(shift)/f90_in3(offset5)
        f90_out(shift)=f90_out(shift)-f90_in2(offset6)
        offset6=offset6+(1)
        offset5=offset5+(1)
        offset4=offset4+(1)
        offset2=offset2+(1)
        offset0=offset0+(1)
        shift=shift+(1)
      END DO
    END DO
  END DO


END SUBROUTINE ADD_SUB3

File: people/thom/test/bak/opsm.c
#include <stdlib.h>
#include "moalib.e"

ADD_SUB1(double _in1[], double _in2[], double _in3[], double _out[])

{
  double *offset0;
  int i0;
  int i1;
  int i2;
  double *shift;

/*******
out=in1
********/

  shift=_out+0*(int) (3)+0*(int) (4)+0;
  offset0=_in1+0*(int) 12+0*(int) 4+0;
  for (i0=0; i0<(2); i0++) {
    for (i1=0; i1<(3); i1++) {
      for (i2=0; i2<(4); i2++) {
        *(shift)= *(offset0);
        offset0+=(1);
        shift++;
      }
    }
  }


/*******
out=out +  + in2
********/

  shift=_out+0*(int) (3)+0*(int) (4)+0;
  offset0=_in2+0*(int) 12+0*(int) 4+0;
  for (i0=0; i0<(2); i0++) {
    for (i1=0; i1<(3); i1++) {
      for (i2=0; i2<(4); i2++) {
        *(shift)+= *(offset0);
        offset0+=(1);
        shift++;
      }
    }
  }


/*******
out=out +  + in3
********/

  shift=_out+0*(int) (3)+0*(int) (4)+0;
  offset0=_in3+0*(int) 12+0*(int) 4+0;
  for (i0=0; i0<(2); i0++) {
    for (i1=0; i1<(3); i1++) {
      for (i2=0; i2<(4); i2++) {
        *(shift)+= *(offset0);
        offset0+=(1);
        shift++;
      }
    }
  }


}


ADD_SUB2(double _in1[], double _in2[], double _in3[], double _out[])

{
  double *offset0;
  int i0;
  int i1;
  int i2;
  double *shift;
  double _temp[2*3*4];

/*******
temp=in1
********/

  shift=_temp+0*(int) (3)+0*(int) (4)+0;
  offset0=_in1+0*(int) 12+0*(int) 4+0;
  for (i0=0; i0<(2); i0++) {
    for (i1=0; i1<(3); i1++) {
      for (i2=0; i2<(4); i2++) {
        *(shift)= *(offset0);
        offset0+=(1);
        shift++;
      }
    }
  }


/*******
temp=temp * in2
********/

  shift=_temp+0*(int) (3)+0*(int) (4)+0;
  offset0=_in2+0*(int) 12+0*(int) 4+0;
  for (i0=0; i0<(2); i0++) {
    for (i1=0; i1<(3); i1++) {
      for (i2=0; i2<(4); i2++) {
        *(shift)*= *(offset0);
        offset0+=(1);
        shift++;
      }
    }
  }


/*******
temp=temp - in3
********/

  shift=_temp+0*(int) (3)+0*(int) (4)+0;
  offset0=_in3+0*(int) 12+0*(int) 4+0;
  for (i0=0; i0<(2); i0++) {
    for (i1=0; i1<(3); i1++) {
      for (i2=0; i2<(4); i2++) {
        *(shift)-= *(offset0);
        offset0+=(1);
        shift++;
      }
    }
  }


/*******
out=temp
********/

  shift=_out+0*(int) (3)+0*(int) (4)+0;
  offset0=_temp+0*(int) 12+0*(int) 4+0;
  for (i0=0; i0<(2); i0++) {
    for (i1=0; i1<(3); i1++) {
      for (i2=0; i2<(4); i2++) {
        *(shift)= *(offset0);
        offset0+=(1);
        shift++;
      }
    }
  }


}


ADD_SUB3(double _in1[], double _in2[], double _in3[], double _out[])

{
  double *offset0;
  double *offset1;
  double *offset2;
  double *offset3;
  double *offset4;
  double *offset5;
  double *offset6;
  int i0;
  int i1;
  int i2;
  double *shift;
  double _temp[2*3*4];

/*******
temp=in2 * in3
********/

  shift=_temp+0*(int) (3)+0*(int) (4)+0;
  offset0=_in2+0*(int) 12+0*(int) 4+0;
  offset1=_in3+0*(int) 12+0*(int) 4+0;
  for (i0=0; i0<(2); i0++) {
    for (i1=0; i1<(3); i1++) {
      for (i2=0; i2<(4); i2++) {
        *(shift)= *(offset0);
        *(shift)*= *(offset1);
        offset1+=(1);
        offset0+=(1);
        shift++;
      }
    }
  }


/*******
out=in1 - four * in3 +  + tfive * temp / in3 - in2
********/

  shift=_out+0*(int) (3)+0*(int) (4)+0;
  offset0=_in1+0*(int) 12+0*(int) 4+0;
  offset2=_in3+0*(int) 12+0*(int) 4+0;
  offset4=_temp+0*(int) 12+0*(int) 4+0;
  offset5=_in3+0*(int) 12+0*(int) 4+0;
  offset6=_in2+0*(int) 12+0*(int) 4+0;
  for (i0=0; i0<(2); i0++) {
    for (i1=0; i1<(3); i1++) {
      for (i2=0; i2<(4); i2++) {
        *(shift)= *(offset0);
        *(shift)-= 4.000000;
        *(shift)*= *(offset2);
        *(shift)+= 25.000000;
        *(shift)*= *(offset4);
        *(shift)/= *(offset5);
        *(shift)-= *(offset6);
        offset6+=(1);
        offset5+=(1);
        offset4+=(1);
        offset2+=(1);
        offset0+=(1);
        shift++;
      }
    }
  }


}


